### PR TITLE
panic on calling close on close channel

### PR DESCRIPTION
**Pull request by syeleti-msft**
Issue: https://github.com/Azure/azure-storage-fuse/issues/1426

Closing the channel twice causing the panic.
This can happen in 3 scenarios:
1. Closing the file.
2. when requesting a block for writing.
3. Evicting the blocks when the file is in rand-read mode.

Both call the same method waitAndFreeUploadedBlocks, which unblocks the requested no of blocks inside the cooked list. but cooked list might contain the blocks which were unblocked already (i.e., when blocks are downloaded they were moved to cooked list when there is a reader for them and unblocked), hence calling unblock on them again causes panic.

### CODE REVIEW

File: component/block_cache/block_cache.go [hunk 1]
```diff
@@ -1213,12 +1213,13 @@ func (bc *BlockCache) waitAndFreeUploadedBlocks(handle *handlemap.Handle, cnt in
...
 			// Wait for upload of this block to complete
-			<-block.state
+			_, ok := <-block.state
+			if ok {
+				block.Unblock()
+			}
 			block.flags.Clear(BlockFlagUploading)
 		}
 
-		block.Unblock()
```
Comment (vibhansa-msft) [inline component/block_cache/block_cache.go:1220]: block-id can be -1 and for those we need to close the channel as well but if id is  != -1 then we shall not wait to receive anything. Hence the code was written like this. block-id == -1 is a valid case where due to prefetch we allocated more blocks then required and we never used them to fill data as there offsets were beyond the file length. So may be better fix is to release those blocks when you try to do lineupDownload() and function returns EOF. That will fix this problem too.

File: component/block_cache/block_cache.go [hunk 2]
```diff
@@ -1098,17 +1101,22 @@ func (bc *BlockCache) getOrCreateBlock(handle *handlemap.Handle, offset uint64)
 			block.flags.Clear(BlockFlagSynced)
 		} else if block.flags.IsSet(BlockFlagDownloading) {
 			log.Debug("BlockCache::getOrCreateBlock : Waiting for download to finish for committed block %v for %v=>%s", block.id, handle.ID, handle.Path)
-			<-block.state
+			_, ok := <-block.state
+			if ok {
+				block.Unblock()
+			}
 			block.flags.Clear(BlockFlagDownloading)
```
Comment (vibhansa-msft) [inline component/block_cache/block_cache.go:1108]: This line we can move up, generally we are doing unblock after all flags are reset.

File: component/block_cache/block_cache.go [hunk 3]
```diff
@@ -1098,17 +1101,22 @@ func (bc *BlockCache) getOrCreateBlock(handle *handlemap.Handle, offset uint64)
...
 		} else if block.flags.IsSet(BlockFlagUploading) {
 			// If the block is being staged, then wait till it is uploaded,
 			// and then write to the same block and move it back to cooking queue
 			log.Debug("BlockCache::getOrCreateBlock : Waiting for the block %v to upload for %v=>%s", block.id, handle.ID, handle.Path)
-			<-block.state
+			_, ok := <-block.state
+			if ok {
+				block.Unblock()
+			}
 			block.flags.Clear(BlockFlagUploading)
```
Comment (vibhansa-msft) [inline component/block_cache/block_cache.go:1118]: same as above: clear flags before the unblock. if ok{ check can be done after the reset

File: component/block_cache/block_cache.go [hunk 4]
```diff
@@ -1234,10 +1242,10 @@ func (bc *BlockCache) waitAndFreeUploadedBlocks(handle *handlemap.Handle, cnt in
 		if block.id != -1 {
 			// Wait for upload of this block to complete
 			_, ok := <-block.state
-			block.flags.Clear(BlockFlagUploading)
```
Comment (vibhansa-msft) [inline component/block_cache/block_cache.go:1237]: This one is the correct way to do it.

**Review by vibhansa-msft — APPROVED** [review]

**Review by souravgupta-msft — APPROVED** [review]

**Pull request merged**